[sjaqjwor(이승기)] step3 구간 추가 기능 PR 요청드립니다.

sjaqjwor opened the pull request:
안녕하세요!

구간 추가 기능 PR 요청드립니다.

감사합니다 😄

sjaqjwor committed:
docs(README) : 요구 사항 분석 및 API 명세

sjaqjwor committed:
refactor : Section 도메인 생성 및 Line에서 Stations를 Sections로 교체

sjaqjwor committed:
test : 노선에 라인을 추가할 수 있는 ATDD 테스트

sjaqjwor committed:
feat : 노선에 section을 추가할 수 있는 기능 구현

sjaqjwor committed:
test : 노선에 구간에 새로운 상행 , 하행 종점을 추가 할 수 있다.

sjaqjwor committed:
feat : 노선 구간에 새로운 상행 , 하행 종점 추가 기능

sjaqjwor committed:
test : 중복된 노선 구간을 저장할 수 없다

sjaqjwor committed:
feat : 중복된 노선 구간을 저장 할 수 없는 기능 구현

sjaqjwor committed:
test : 이미 있는 노선 사이에 새로운 구간을 저장 할 수 있다.

sjaqjwor committed:
feat : 이미 있는 노선 사이에 새로운 구간 저장하는 기능

sjaqjwor committed:
test : 새로운 구간을 저장하려고 하지만 기존 구간의 길이보다 길거나 동일해서 저장 할 수 없다.

sjaqjwor committed:
feat : 새로운 구간 등록 시 distance 검사하는 기능 추가

sjaqjwor committed:
test: 새로운 구간 등록시 upstation, downstation이 모두 기존 구간에 들어 있지 않아 저장 실패

sjaqjwor committed:
feat : 구간 등록 시 전혀 없는 상행 , 하행역이 들어온 경우 저장 실패 기능 추가

sjaqjwor committed:
refactor : 이미 있는 구간인지 확인할 때 upStation ,downStation 바꿔서도 확인

sjaqjwor committed:
refactor : 구간 중간 저장 시 실패 나는 케이스 리팩토링

sjaqjwor committed:
refactor : 오타수정

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -23,12 +23,12 @@
 @DisplayName("지하철노선 관련 기능")
 public class LineAcceptanceTest extends AcceptanceTest {
-    private static final String LINE_MAIN_PATH = "/lines";
+    protected static final String LINE_MAIN_PATH = "/lines";
```
다른 테스트에서 함께 쓰기 위해 상수의 접근 제어자를 변경하신 것일까요?
테스트에서 공통으로 쓰기 위한 상수는 별도 클래스를 만들어서 분리해보셔도 좋겠습니다. :)
테스트 파일 내에 있는 상수를 다른 테스트에서도 사용하면, 상수 내용이 변경되었을 때 사용하는 모든 테스트를 확인해야 하는데, 이를 개발하는 단계에서 놓치기 쉬울 것 같습니다.

sjaqjwor replied:
새롭게 LineSectionAcceptanceTest를 생성했는데

LineAcceptanceTest에 중복된것들이 있어서 최초에는 해당 클래스를 상속해서 쓰려고 했으나

그렇게 되면 테스트코드가 LineAcceptanceTest까지 돌기 때문에 좋지 않다고 판단하여 

다시 원상태로 되돌렸는데... 저부분은 수정하지 못했네요

공통된것들을 추출해서 하는 방향으로 수정해보도록 하겠습니다.

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,157 @@
+        assertThat(stations).hasSize(3);
+    }
+
+    @DisplayName("이미 존재하는 구간 사이에 새로운 구간을 등록 하려고 하나 기존 구간의 길이보다 크거나 , 같아서 저장 할 수 없다.")
```
크거나 같은 경우에 대한 검증을 모두 해보시면 좋을 것 같습니다.
`@ParameterizedTest` 를 사용해보면 어떨까요?

sjaqjwor replied:
넵넵

검증을 좀더 세분화해서 수정해 보도록 하겠습니다.

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,157 @@
+import static nextstep.subway.line.LineAcceptanceTest.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("노선 구간 관련 기능")
```
인수 테스트 잘 작성해주셨습니다. 👍 

인수 테스트간의 의존성에 대해서도 고민해보시면 좋겠습니다.
하나의 테스트 안에 있는 메서드는 보통 그 테스트 안에서만 사용되는데, 다른 테스트에서도 함께 사용한다면 메서드가 변경될 때 사용하는 테스트를 모두 찾아서 확인해야 한다는 것을 개발자가 기억해야 할 것 같아요.

테스트간에 공통으로 사용하는 메서드는 유틸리티로 만들어 보시는 것도 좋을 것 같네요!

sjaqjwor replied:
지금 현재 공통으로 사용되는 findById는 공통으로 사용하지 않게 될꺼 같습니다.

그 이유는 해당 findById메소드로 검색을 하게 되면 distance를 비교 검증을 할 수 없기 때문에

상수 영역만 공통으로 사용되는 것을 추출하고

해당 메소드는 제거 후 다른 controller을 사용하여 distance를 검증할 생각입니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.stream.Collectors;
+
+@Embeddable
+public class Sections {
```
일급 컬렉션을 만들고 구간에 대한 관리를 이 클래스에서 하도록 구현하신 것 좋습니다. 👍

sjaqjwor replied:
감사합니다 😄

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,105 @@
+        return upStation;
+    }
+
+    public void validateDuplicate(Section section) {
```
`Section` 은 다른 `Section` 과의 관계를 관리하는 클래스가 아니므로, 중복인지에 대한 확인은 여기에서 하고 중복이 있을 경우 예외를 발생시키는 것은 `Sections` 클래스에서 하면 어떨까요?

sjaqjwor replied:
처음에 생각했던건 결국 section과 section을 비교하는것이기 때문에 중복 검사를

section에서 하고 해당 메소드에서 발생시키는게 맞지 않나라고 생각하게 되었습니다.

@seokhongkim 님께서는 말씀하신것 처럼

사실 Sections 클래스에서 section을 관리하기 때문에 

단지 동일한 여부만 section이 판단하고

그에대한 exception은 Sections 클래스에서 하는게 좀 더 타당하다는 생각이 듭니다.

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,157 @@
+
+    @DisplayName("이미 존재하는 구간 사이에 새로운 구간을 등록 할 수 있다.")
+    @Test
+    void addSectionMiddle() {
```
구간 중간에 역을 추가하는 경우, 거리 계산이 잘 되었는지도 검증해보시면 좋겠습니다.

sjaqjwor replied:
넵넵

해당 부분도 검증을 좀더 구체적으로 해보겠습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,124 @@
+    @OneToMany(mappedBy = "line", cascade = CascadeType.ALL)
+    private List<Section> values = new ArrayList<>();
+
+    public void addSection(Line line, Station upStation, Station downStation, Long distance) {
```
승기님께서는 메서드를 배치할 때 어떤 순서로 배치하시나요?

sjaqjwor replied:
메서드 배치는 생각해 본적이 크게 없는것 같습니다...

피드백 반영하면서 든 생각은 

A라는 public 메소드가 있다면 A라는 메소드의 결과를 만들기 위한

메소드들끼리 뭉쳐있으면 좋겠다는 생각이 들었습니다.

혹시 @seokhongkim  어떻게 배치 하실까요?

seokhongkim replied:
@sjaqjwor 승기님,
저는 메서드 중에서는 public 메서드를 먼저 작성하고, 해당 메서드 내부에서 호출하는 private 메서드를 사용되는 순서대로 이어서 작성합니다.  
오버로딩된 public 메서드가 존재하는 경우, 해당 메서드는 모아서 작성하고, 내부에서 호출하는 private 메서드를 사용되는 순서대로 이어서 작성합니다.
이렇게 이어서 작성했을 때, public 메서드 내부에서 사용하는 메서드를 가까운 곳에서 찾을 수 있고, 순서대로 찾아볼 수 있어서 구현된 코드를 읽기 위해 이동하는 시간을 줄일 수 있어 좋았습니다.

관련하여 클린 코드에 나온 내용을 참고해보셔도 좋을 것 같습니다. :)

sjaqjwor replied:
@seokhongkim 

넵넵

이번에 구현 할때는 참고 해서 적용해보도록 하겠습니다.

감사드립니다 😄

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,124 @@
+        addSection(section);
+    }
+
+    private void validateSectionExist(Section section) {
```
구간을 추가했을 때 노선에 있는 역이 하나도 없는 경우를 검증하는 것 같네요.
메서드의 이름과 메서드에서 사용하는 변수의 이름이 기능을 조금 더 잘 설명해줄 수 있으면 좋겠습니다. :)

sjaqjwor replied:
넵넵 

알겠습니다!!

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    private boolean isAddDownStationTerminus(Section section) {
+        return this.values.stream().anyMatch(v -> v.getDownStation().equals(section.getUpStation()));
```
[1, 2], [2, 3], [3, 4]로 이루어진 구간이 있을 경우(1 - 2 - 3 - 4), 현재 코드의 로직에 따르면 [3, 10] 이라는 구간을 추가했을 때(1 - 2 - 3 - 10 - 4)도 이 조건을 만족해서 [3, 10] 구간만 추가할 것으로 보이는데요. 의도하신 부분이 맞을까요?

sjaqjwor replied:
넵넵 해당 구간을 추가 하고

기존에 있던 3-4에 구간을 3-10 추가 , 10-4로 변경하게 됩니다.

혹시 문제가 있을까요??

sjaqjwor replied:
@seokhongkim 
이 부분은 뭔가 제가 잘 못 구현한것 같습니다..

인수테스트 추가 해서 재 구현 하도록 하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,124 @@
+                .filter(v -> !downStations.contains(v)).findFirst().orElseThrow(NotFoundStation::new);
+    }
+
+    private List<Station> createStations(Station upStationTerminus) {
```
중간에 빈 줄을 사용하여 코드를 나눠주시면 가독성이 좋아질 것 같습니다. :)

```
// 지하철 역 목록을 만든다.
(optional empty line)
// 상행 종점을 목록에 추가한다.

// 하행 역 검색을 위해 필요한 맵을 만든다.
// 첫 번째 하행 역을 가져온다.
// 구간을 탐색하며 하행 역을 목록에 추가한다.

// 전체 역 목록을 반환한다.
```

sjaqjwor replied:
넵넵 감사합니다.

혹시 주석도 필요로 할까요? 주석까지는 불필요 할꺼 같다는 생각이 들어서 여쭤봅니다.,.ㅠ

seokhongkim replied:
아뇨, 주석은 필요 없습니다. :)
코드를 쓰는 것보다 코드의 설명을 한글로 썼을 때 가독성 비교가 더 쉽다고 생각해서, 코드를 한글로 풀어서 쓴 내용입니다.

seokhongkim reviewed (changes requested):
승기님, 3단계 미션 잘 구현해주셨습니다. 👍 

코드를 보며 궁금한 점과 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

sjaqjwor committed:
refactor(피드백) : 공통 부분으로 사용되는 constant 분리 및 공통 사용 메소드 제거

sjaqjwor committed:
refactor(피드백) : distance비교를 위한 리팩토링(자세한 내용 아래 커밋 내용 참조)

update:

1. LineSectionAcceptanceTest에서 사용하던 findById 제거
2. 라인에 해당하는 section을 반환하는 컨트롤러 생성
3. 해당 컨트롤러를 통해 section정보를 가져와서 distance를 비교
4. section 정보를 반환하는 response 생성
5. 정렬된 section을 가져오는 메소드 생성

sjaqjwor committed:
refactor(피드백) : Station을 정렬하는 부분 라인 쉽게 보기 위해 라인 정리

sjaqjwor committed:
refactor(피드백) : addSectionMiddleFailedDistance에서 길거나 , 동일한 distance 테스트 코드 추가(ParameterizedTest))

sjaqjwor committed:
refactor(피드백) : section에 있던 중복 exception제거 및 sections에서 excpetion throw하도록 변경

sjaqjwor committed:
refactor(피드백) : 메소드 및 변수명 변경

sjaqjwor committed:
test : 이미 많은 구간이 존재할 때 중간에 구간 인서트 테스트

sjaqjwor committed:
refatcor : 이미 많은 구간이 존재할 때 구간 인서트 시 에러 발생하여 해당 기능 수정

sjaqjwor committed:
refactor : 정렬된 모든 지하철역 , 정렬된 모든 구간 Method명 변경

sjaqjwor committed:
refactor(피드백) : 메소드 순서 기능 단위로 순서 변경

sjaqjwor commented:
@seokhongkim 

피드백 주신거 반영했습니다.

메소드 순서 같은 경우에는

기능별로 한번 정렬해봤는데...

어떤 방식이 좋을지 의견 부탁드립니다!!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -64,4 +61,9 @@ public LineResponse createSection(Long lineId, SectionSaveRequest sectionRequest
         return LineResponse.of(line);
     }
+
+    public SectionsResponse findSectionByLine(Long lineId) {
```
`findSectionsByLine`(Section 뒤에 s 추가)로 이름을 지어보면 어떨까요? :)
응답 타입을 보고 목록을 조회하는 것임을 알 수 있지만, 메서드 이름도 복수형으로 작성해주면 더 명시적인 것 같아요!

sjaqjwor replied:
@seokhongkim 

넵넵

삭제 과제할때 반영하도록 하겠습니다.

감사합니다. 😄

seokhongkim reviewed (approved):
승기님, 3단계 미션 코멘트 잘 반영해주셨습니다. 👍 
3단계 PR 머지하도록 하겠습니다.

메서드 순서와 관련해서는 질문주신 코멘트에 답변으로 남겨 놓았으니 참고 부탁드립니다. 🙏 
4단계 미션에서 뵙겠습니다. 😀
